File: Programmers/2019카카오겨울인턴십/호텔방배정/호텔방배정.py
rooms = {}

def solution(k, room_number):
    reserved = [False] * (k+1)
    
    def assign_rooms(number):
        if not reserved[number]:
            reserved[number] = True
            rooms[number] = number+1
            return number

        near_number = assign_rooms(rooms[number])
        rooms[near_number] = near_number+1
        return near_number
             
    return list(map(assign_rooms, room_number))
        

sol = solution(10, [1,3,4,1,3,1])
print(sol)

File: Programmers/2019카카오겨울인턴십/호텔방배정/호텔방배정.js
const solution = (k, room_number) => {
  const rooms = new Array(k+1);
  
  const answer = room_number.map((num) => {
    if(!rooms[num]) {
      rooms[num] = true;
      return num;
    }
    
    for(let i=num+1; i<k+1; i++) {
      if(!rooms[i]) {
        rooms[i] = true;
        return i;
      }
    }
  })

  return answer;
};

const sol = solution(10, [1, 3, 4, 1, 3, 1]);
console.log(sol); //[1,3,4,2,5,6]
